### DISCUSSION TITLE
Name for UUID with all bits set to 1 (FFFFFFFF-FFFF-FFFF-FFFF-FFFFFFFFFFFF)

### DISCUSSION THREAD

**Discussion by kyzer-davis** [uuid6/prototypes#46] [Polls]
Feel free to share the poll. I will close this on June 9th at 9am EST as to ensure I can update the doc for IETF Working Group Last Call (WGLC)

---

See discussions below for full context as to the choices:
- https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/107
- https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/95
- https://github.com/uuid6/uuid6-ietf-draft/issues/62

Edit: All bits set to 0 remains "Nil UUID" as per the old definition defined by RFC4122.

### Samples
![image](https://github.com/uuid6/prototypes/assets/867657/3f880451-94e5-4a09-9522-b38ccbeea6ed)

![image](https://github.com/uuid6/prototypes/assets/867657/3cfe123f-035e-489d-a954-f1cd3bc0e7b3)

---

CC: @jdknezek, @daegalus, @chrylis, @mikemix, @kurttheviking, @sprql, @kjmph, @MatrixAI, @kripod, @karwa, @mareek, @oittaa, @jakwings, @fabiolimace, @LiosK, @theckman @kjmph @iology @BaerMitUmlaut @DianaNites @jakwings @medo64 @nicolas-grekas @mlhamel @tgross35 @ben221199 @bradleypeabody @mcr @martinthomson @jimfenton @petere @astiob @DianaNites @crutchcorn @broofa

**Comment by BaerMitUmlaut** [comment]
Have you considered NaU - Not a UUID? A float with all bits set to 1 is NaN, and since this would be an invalid version, it would not be a valid UUID.

**Reply by medo64** [reply]
While that wouldn't be a valid UUID now, there is a small risk that we'll have some future version that would make it valid...
And it also opens a question if other invalid bit patterns should be also matched, e.g. should the following check in programming language work (analog for NaN behaviour in double):
```
if (UUID.IsNaU(0xFFFFFFFF-FFFF-FFFF-FFFF-FFFFFFFFFFFF))
if (UUID.IsNaU(0xFFFFFFFF-FFFF-FFFF-FFFF-FFFFFFFFFFFE))
if (UUID.IsNaU(0xFFFFFFFF-FFFF-FFFF-FFFF-FFFFFFFFFFF0))
...
```
All three are essentially NaU based on their version bits but only the first one is `Max`...

**Reply by astiob** [reply]
~~In that sense, all-zeros is not a valid UUID and would be eligible for that name as well, right?~~ Never mind, I missed the mention of version. Sorry for the noise.

**Reply by ben221199** [reply]
@astiob The `Nil UUID` falls in the range of the legacy Apollo NCS UUID variant. When looking at the families in that variant, you see that the `Nil UUID` is from the family `AF_UNSPEC`, so the format is unspecified and totally valid to be all zeros.

**Reply by kyzer-davis** [reply]
It would be valid just as much as Nil UUID as we are specially allocating it in the spec and citing it in all places.
- All 0s = Nil UUID from RFC4122 carried into current draft.
- All 1s = To be Named UUID from out latest draft.

**Comment by ben221199** [comment]
I wrote a whole text, but my laptop got a BSOD. Text gone. Big F.

---

**What I wanted to say was:** I don't think it is the best option to only make a poll between the proposed "Omni" option and the old "Max" option. It is possible that there is a better option out there besides those two that fits more. I tried to find some more (some with the help of ChatGPT):

- Not Nil UUID
- Anti Nil UUID
- Lin UUID (`Lin` is mirror of `Nil`)
- All-ones UUID
- All-high UUID
- Full UUID
- Unity UUID ("Uno" is "one")
- Solid UUID
- Saturated UUID
- Pinnacle UUID
- Top UUID

You can read the chat here: https://chat.openai.com/share/97dbf284-129f-420b-b013-cd2b8f6a62f0. Somehow it also came up with both "Max" UUID and "Omni" UUID.

**Reply by kyzer-davis** [reply]
> Somehow it also came up with both "Max" UUID and "Omni" UUID.

Awesome.

**Comment by daegalus** [comment]
So, I still don't get why people are so upset about the naming. NIL UUID is in the original RFC, and its NIL because its all 0s. 0 is NIL, done. 

For the new Max UUID, it makes sense. All F is the maximum possible value of UUID. I don't think any other UUID makes sense. UUID isn't a mask, it is a value. a 128bit value encoded into hexadecimal. So it has a max value.

If there is any change, I would prpose NIL be renamed MIN, and then we have MIN/MAX which equally makes sense for the UUID value.

**Reply by fabiolimace** [reply]
I think the controversy started when we put too much emphasis on the etymology of "nil", but it's not a matter of philology. Most likely, the original authors just meant ZERO UUID. Maybe it's not a controversy for native English speakers, which I'm not. Please correct me if I'm wrong.

https://en.wikipedia.org/wiki/Nil
https://www.merriam-webster.com/dictionary/nil
https://dictionary.cambridge.org/dictionary/english/nil
https://en.wikipedia.org/wiki/Names_for_the_number_0_in_English

**Reply by daegalus** [reply]
Maybe. I speak 4 languages, and Nil means Zero in all of them, or it has no meaning outside of computer science.

I always interpreted NIL UUID as ZERO UUID. All zeros, means the value is 0, hence NIL, nothing, nada.

Omni to me means Everything, but all F UUID isnt everything, it doesn't encompass a range, or a totality, it just means maximum value, since thats all it is. the maximum value of 128bits. which is a very large number, but still a value. We even treat UUIDs as values, not as ranges, or masks, or anything of the sort.

Hence why Omni is very confusing to me and makes no sense.

so ZERO UUID or MIN UUID makes the most sense to me if there was any change to these 2 values.

**Reply by kyzer-davis** [reply]
Agreed, if I was writing the spec from scratch I would have probably opted for min/max.
...but at this point Nil has been around for a few decades or so. It is found in lot's of documentation, code snippets/comments, and who knows where else. I would rather leave the all 0s UUID to the name that is currently ubiquitous around the net.

**Reply by ben221199** [reply]
It would be also possible to name it "zero". However, "min" would be dependent on being signed and unsigned and a 128 bit number could be both. This is also the case for "max". The "nil" and "zero" are always "nil" and "zero", both signed and unsigned.

**Reply by astiob** [reply]
My impression is that the main motivation for a nil UUID has nothing to do with ordering, which is/would be implied by min/max: it’s just the UUID to use when you have no (meaningful) UUID; and all-zeros are merely a convenient choice for a number of reasons including psychology, zero-initialization and, yes, the fact that it is not in the middle of any range of meaningful UUIDs.

When one does order UUIDs in a sequence, I think it’s unequivocal that they aren’t signed integers and not quite integers at all, so signed/unsigned doesn’t really apply.

It seems the main motivation for the all-ones UUID _is_ ordering, so “max UUID” seems more pertinent in that sense. (And it does imply that UUIDs, if treated as integers, are unsigned. Or if treated as multiple integers each, that all the parts are unsigned.)

Personally, my first instinct was to vote for “max” because “omni” just sounded weird, like it somehow _enveloped_ other UUIDs, but I do slightly dislike how it introduces extra ambiguity between “the biggest of the UUIDs we have in a particular given list” and “the biggest of all UUIDs that could possibly ever exist”. Ultimately, I think either name works and we’ll get used to it.

**Comment by chrylis** [comment]
I mostly live in JVM-land, and I'm used to seeing `MAX` in analogous contexts (as in Java `Integer.MAX_VALUE` or C `INT_MAX`). I'd already adopted the name "max UUID" informally and think it's the clearest name. (I'd also have preferred the name "zero UUID" to "nil", but that's water under the bridge.)

**Comment by mcr** [comment]
How about "Omni-F" :-)
MAX UUID is probably more understandable though.
I also don't mind if we just have no term.

**Comment by BaerMitUmlaut** [comment]
Another suggestion: LIMIT
Since MAX sort of implies there to be a MIN, this might be a better choice.

**Comment by LiosK** [comment]
Was wondering if we can say something nerdy like: there exist Max, Nil (zero), and Min (negative F...F) UUIDs, but we can't perceive Min because it's below the horizon/in an imaginary world/compactified in hidden dimensions.

**Reply by ben221199** [reply]
You mean that we say it is a 129-bit number with the sign-bit removed to make it 128-bit. 😂
